put focus on primary focus element in modal if specified

teresamonahan opened the pull request:
Closes IBM/carbon-components-angular#102

Allows the developer to specify what element within a modal initial focus should go to when the modal is opened.

#### Changelog

**Changed**

* selectorPrimaryFocus property added to the Modal which determines which element to focus.
* ngOnInit updated to find the element with the selectorPrimaryFocus selector and focus is called on it if available

Behavior after this change when the `data-modal-primary-focus` selector is placed on the primary button in the modal footer - note the focus styling on the close button in the modal footer, instead of the previous blue border on the entire modal:

![modal-focus](https://user-images.githubusercontent.com/2227544/45891328-33c7c980-bdbd-11e8-89ad-65cbe4e41c04.gif)

cal-smith commented on src/modal/modal.component.ts:
```diff
@@ -148,6 +148,12 @@ export class ModalComponent implements OnInit, OnDestroy {
+	 * An element should have 'data-modal-primary-focus' as an attribute to receive initial focus within the `Modal` component.
+	 * @memberof ModalComponent
+	 */
+	selectorPrimaryFocus = "[data-modal-primary-focus]";
```
since this all passes through the angular template parser we should be okay with just `[modal-primary-focus]` ... thoughts?

asudoh replied:
Weighing in - I guess it's an attempt to align to our vanilla code, though such alignment is not a hard requirement.

teresamonahan replied:
Yes I see this selector is used in both vanilla and react code so thought it best to use the same here to make it easier for people transitioning between the different implementations.

zvonimirfras replied:
We don't use the `data-` prefix anywhere else unless required by the carbon css.

cal-smith commented on src/modal/modal.component.ts:
```diff
@@ -161,6 +167,11 @@ export class ModalComponent implements OnInit, OnDestroy {
+		if (primaryFocusElement && primaryFocusElement.focus) {
+			primaryFocusElement.focus();
+			return;
+		}
```
this should be run in an `ngAfterViewInit` since the view isn't guaranteed to exist until then.

teresamonahan commented:
@cal-smith @asudoh Comments addressed. Can you review again please?

zvonimirfras commented on src/modal/modal.component.ts:
```diff
@@ -156,11 +163,23 @@ export class ModalComponent implements OnInit, OnDestroy {
+	 * @memberof ModalComponent
+	 */
+	ngAfterViewInit() {
+		const primaryFocusElement  = this.modal.nativeElement.querySelector(this.selectorPrimaryFocus);
```
There's an extra space before `=`

zvonimirfras commented on src/modal/modal.component.ts:
```diff
@@ -156,11 +163,23 @@ export class ModalComponent implements OnInit, OnDestroy {
+			primaryFocusElement.focus();
+			return;
+		}
 		this.modal.nativeElement.focus();
```
While we're on this topic, when modal is opened, focus goes to first focusable element for accessibility. That code would go here instead focusing the modal element.

teresamonahan commented:
@zvonimirfras I updated the selector to remove 'data-' and removed the extra space.
I created a new issue focusing the first focusable element by default as I don't have time to work on that right now - https://github.com/IBM/carbon-components-angular/issues/124
